osquery/tables/system/darwin/signature.mm

```diff
@@ -58,130 +62,159 @@ Status getVerifyFlags(SecCSFlags& flags) {
   }
 
-  // Create the static code object.
-  SecStaticCodeRef staticCode = nullptr;
-  result = SecStaticCodeCreateWithPath(url, kSecCSDefaultFlags, &staticCode);
-  if (result != errSecSuccess || staticCode == nullptr) {
-    VLOG(1) << "Could not create static code object for file: " << path;
-    return;
+  if (arch.empty()) {
+    // Create the static code object.
+    result = SecStaticCodeCreateWithPath(url, kSecCSDefaultFlags, &static_code);
+    if (result != errSecSuccess) {
+      if (static_code != nullptr) {
+        CFRelease(static_code);
+      }
+      return Status(1, "Could not create static code object");
+    }
+  } else {
+    CFMutableDictionaryRef context =
+        CFDictionaryCreateMutable(nullptr,
+                                  0,
+                                  &kCFTypeDictionaryKeyCallBacks,
+                                  &kCFTypeDictionaryValueCallBacks);
+    auto cfkey = CFStringCreateWithCString(
+        kCFAllocatorDefault, arch.c_str(), kCFStringEncodingUTF8);
+    CFDictionaryAddValue(context, kSecCodeAttributeArchitecture, cfkey);
+    CFRelease(cfkey);
+    result = SecStaticCodeCreateWithPathAndAttributes(
+        url, kSecCSDefaultFlags, context, &static_code);
+    CFRelease(context);
+    if (result != errSecSuccess) {
+      if (static_code != nullptr) {
+        CFRelease(static_code);
+      }
+      return Status(0, "No code to verify");
+    }
   }
 
-  // Actually validate.
-  bool isValidated = false;
-  result =
-      SecStaticCodeCheckValidityWithErrors(staticCode, flags, nullptr, nullptr);
+  Row r;
+  r["path"] = path;
+  r["arch"] = arch;
+  r["identifier"] = "";
+
+  SecCSFlags flags = 0;
+  getVerifyFlags(flags);
+  result = SecStaticCodeCheckValidityWithErrors(
+      static_code, flags, nullptr, nullptr);
   if (result == errSecSuccess) {
-    isValidated = true;
+    r["signed"] = "1";
   } else {
     // If this errors, then we either don't have a signature, or it's malformed.
-    VLOG(1) << "Static code validity check failed for file: " << path;
+    r["signed"] = "0";
   }
-  CFDictionaryRef codeInfo = nullptr;
+
+  CFDictionaryRef code_info = nullptr;
   result = SecCodeCopySigningInformation(
-      staticCode, kSecCSSigningInformation | kSecCSRequirementInformation,
-      &codeInfo);
-  if (result == errSecSuccess) {
-    // If we don't get an identifier for this file, then it's not signed.
-    CFStringRef ident =
-        (CFStringRef)CFDictionaryGetValue(codeInfo, kSecCodeInfoIdentifier);
-    if (ident != nullptr) {
-      // We have an identifier - this indicates that the file is signed,
-      // and, since it didn't error above, it's *also* a valid signature.
-      if (isValidated) {
-        r["signed"] = INTEGER(1);
-      }
-      r["identifier"] = stringFromCFString(ident);
+      static_code,
+      kSecCSSigningInformation | kSecCSRequirementInformation,
+      &code_info);
+
+  if (result != errSecSuccess) {
```

Comment:
Is `code_info` guaranteed to be null when the function fails? The documentation is not explicit about it.

Reply:
We can add an extra check here.